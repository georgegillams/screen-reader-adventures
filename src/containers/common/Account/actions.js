import { createRoutine } from 'redux-saga-routines';

import { LOGOUT, REQUEST_VERIFICATION_EMAIL } from './constants';

export const logout = createRoutine(LOGOUT);
export const requestVerificationEmail = createRoutine(REQUEST_VERIFICATION_EMAIL);
