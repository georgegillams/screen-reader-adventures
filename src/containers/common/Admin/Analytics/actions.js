import { createRoutine } from 'redux-saga-routines';

import { LOAD } from './constants';

export const load = createRoutine(LOAD);
