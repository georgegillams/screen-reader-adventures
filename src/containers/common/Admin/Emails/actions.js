import { createRoutine } from 'redux-saga-routines';

import { LOAD, RESEND } from './constants';

export const load = createRoutine(LOAD);
export const resend = createRoutine(RESEND);
