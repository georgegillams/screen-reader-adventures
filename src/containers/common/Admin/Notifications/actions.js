import { createRoutine } from 'redux-saga-routines';

import { LOAD, CREATE, REMOVE } from './constants';

export const load = createRoutine(LOAD);
export const create = createRoutine(CREATE);
export const remove = createRoutine(REMOVE);
