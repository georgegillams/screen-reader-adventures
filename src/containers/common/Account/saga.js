import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { logout, requestVerificationEmail } from './actions';
import { setUser } from 'containers/common/Authenticator/actions';

export function* doLogout() {
  const requestURL = apiStructure.logout.fullPath;

  try {
    yield put(logout.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
    });

    if (result.error) {
      yield put(logout.failure(result));
    } else {
      yield put(logout.success(result));
      yield put(setUser.trigger(null));
    }
  } catch (err) {
    yield put(logout.failure(err));
  } finally {
    yield put(logout.fulfill());
  }
}

export function* doRequestVerificationEmail() {
  const requestURL = apiStructure.requestVerificationEmail.fullPath;

  try {
    yield put(requestVerificationEmail.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
    });

    if (result.error) {
      yield put(requestVerificationEmail.failure(result));
    } else {
      yield put(requestVerificationEmail.success(result));
    }
  } catch (err) {
    yield put(requestVerificationEmail.failure(err));
  } finally {
    yield put(requestVerificationEmail.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(logout.TRIGGER, doLogout);
  yield takeLatest(requestVerificationEmail.TRIGGER, doRequestVerificationEmail);
}
