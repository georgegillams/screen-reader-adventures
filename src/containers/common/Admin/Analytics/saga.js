import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { load } from './actions';

export function* doLoad() {
  const requestURL = apiStructure.loadAnalyticsSummary.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.analytics));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
}
