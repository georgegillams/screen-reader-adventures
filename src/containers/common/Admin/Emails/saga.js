import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { load, resend } from './actions';
import { selectState } from './selectors';

export function* doLoad() {
  const requestURL = apiStructure.loadEmails.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.emails));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export function* doResend() {
  const currentState = yield select(selectState());
  const { emailToResend } = currentState;
  const requestURL = apiStructure.resendEmail.fullPath;

  try {
    yield put(resend.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(emailToResend),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(resend.failure(result));
    } else {
      yield put(resend.success(result.emails));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(resend.failure(err));
  } finally {
    yield put(resend.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
  yield takeLatest(resend.TRIGGER, doResend);
}
