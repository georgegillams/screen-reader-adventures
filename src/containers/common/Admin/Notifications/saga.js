import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { selectState } from './selectors';
import { load, create, remove } from './actions';

export function* doLoad() {
  const requestURL = apiStructure.loadNotifications.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.notifications));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export function* doCreate() {
  const currentState = yield select(selectState());
  const { notificationToCreate } = currentState;
  const requestURL = apiStructure.createNotification.fullPath;

  try {
    yield put(create.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(notificationToCreate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(create.failure(result));
    } else {
      yield put(create.success(result));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(create.failure(err));
  } finally {
    yield put(create.fulfill());
  }
}

export function* doRemove() {
  const currentState = yield select(selectState());
  const { notificationToRemove } = currentState;
  const requestURL = apiStructure.deleteNotification.fullPath;

  try {
    yield put(remove.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(notificationToRemove),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(remove.failure(result));
    } else {
      yield put(remove.success(result));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(remove.failure(err));
  } finally {
    yield put(remove.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
  yield takeLatest(create.TRIGGER, doCreate);
  yield takeLatest(remove.TRIGGER, doRemove);
}
