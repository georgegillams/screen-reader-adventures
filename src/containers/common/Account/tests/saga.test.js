import { put, takeLatest } from 'redux-saga/effects';

import { logout, requestVerificationEmail } from '../actions';

import saga, { doLogout, doRequestVerificationEmail } from '../saga';

describe('Account saga', () => {
  let mainSaga;

  beforeEach(() => {
    mainSaga = saga();
  });

  it('Should logout on logout TRIGGER', () => {
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(logout.TRIGGER, doLogout));
  });

  it('Should request email verification on requestEmailVerification TRIGGER', () => {
    mainSaga.next();
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(requestVerificationEmail.TRIGGER, doRequestVerificationEmail));
  });

  describe('logout actions', () => {
    let logoutGenerator;

    const response = {
      success: 'you are logged out',
      status: 200,
    };

    beforeEach(() => {
      logoutGenerator = doLogout();

      const selectDescriptor = logoutGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call logout.success on successful API call', () => {
      logoutGenerator.next();
      const putSuccess = logoutGenerator.next(response).value;
      logoutGenerator.next();

      expect(putSuccess).toEqual(put(logout.success(response)));
    });

    it('Should call logout.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      logoutGenerator.next();
      const putSuccess = logoutGenerator.next(response).value;
      logoutGenerator.next();

      expect(putSuccess).toEqual(put(logout.failure(response)));
    });

    it('Should call logout.failure if an exception occurs', () => {
      const response = new Error('Some error');
      const putFailure = logoutGenerator.throw(response).value;

      expect(putFailure).toEqual(put(logout.failure(response)));
    });
  });

  describe('requestVerificationEmail actions', () => {
    let requestVerificationEmailGenerator;

    const response = {
      success: 'Verification email resent',
      status: 200,
    };

    beforeEach(() => {
      requestVerificationEmailGenerator = doRequestVerificationEmail();

      const selectDescriptor = requestVerificationEmailGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call requestVerificationEmail.success on successful API call', () => {
      requestVerificationEmailGenerator.next();
      const putSuccess = requestVerificationEmailGenerator.next(response).value;
      requestVerificationEmailGenerator.next();

      expect(putSuccess).toEqual(put(requestVerificationEmail.success(response)));
    });

    it('Should call requestVerificationEmail.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      requestVerificationEmailGenerator.next();
      const putSuccess = requestVerificationEmailGenerator.next(response).value;
      requestVerificationEmailGenerator.next();

      expect(putSuccess).toEqual(put(requestVerificationEmail.failure(response)));
    });

    it('Should call requestVerificationEmail.failure if an exception occurs', () => {
      const response = new Error('Some error');
      const putFailure = requestVerificationEmailGenerator.throw(response).value;

      expect(putFailure).toEqual(put(requestVerificationEmail.failure(response)));
    });
  });
});
