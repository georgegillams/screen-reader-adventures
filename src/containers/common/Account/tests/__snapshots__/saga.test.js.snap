// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account saga logout actions Should call logout.failure if API returns error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/account/LOGOUT/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`Account saga logout actions Should call logout.failure if an exception occurs 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/account/LOGOUT/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`Account saga logout actions Should call logout.success on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/account/LOGOUT/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`Account saga requestVerificationEmail actions Should call requestVerificationEmail.failure if API returns error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/account/REQUEST_VERIFICATION_EMAIL/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`Account saga requestVerificationEmail actions Should call requestVerificationEmail.failure if an exception occurs 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/account/REQUEST_VERIFICATION_EMAIL/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`Account saga requestVerificationEmail actions Should call requestVerificationEmail.success on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/account/REQUEST_VERIFICATION_EMAIL/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
